File: CartQuestiOS/CartQuestiOS/Views/Auth/LoginViewModel.swift
import Foundation
import FirebaseAuth

enum AuthState {
    case unauthenticated
    case loading
    case authenticated(FirebaseAuth.User)
    case error(String)
}

@Observable
class LoginViewModel {
    var authState: AuthState = .unauthenticated
    var email: String = ""
    var password: String = ""
    var displayName: String = ""
    var isSignUp: Bool = false

    private let userRepository = UserRepository()

    init() {
        if let user = Auth.auth().currentUser {
            authState = .authenticated(user)
        }
    }

    func signInWithEmail() {
        authState = .loading
        Task {
            do {
                let result = try await Auth.auth().signIn(withEmail: email, password: password)
                await MainActor.run {
                    authState = .authenticated(result.user)
                }
            } catch {
                await MainActor.run {
                    authState = .error(error.localizedDescription)
                }
            }
        }
    }

    func signUpWithEmail() {
        authState = .loading
        Task {
            do {
                let result = try await Auth.auth().createUser(withEmail: email, password: password)
                let changeRequest = result.user.createProfileChangeRequest()
                changeRequest.displayName = displayName
                try await changeRequest.commitChanges()
                try await userRepository.createUserDocument(user: result.user)
                await MainActor.run {
                    authState = .authenticated(result.user)
                }
            } catch {
                await MainActor.run {
                    authState = .error(error.localizedDescription)
                }
            }
        }
    }

    func signInWithGoogle() {
        // TODO: Implement Google Sign-In once GoogleSignIn-iOS SPM package is added
        // Will use GIDSignIn to get ID token, then:
        // let credential = GoogleAuthProvider.credential(withIDToken: idToken, accessToken: accessToken)
        // try await Auth.auth().signIn(with: credential)
    }

    func signOut() {
        try? Auth.auth().signOut()
        authState = .unauthenticated
    }
}

File: CartQuestAndroid/app/src/main/java/com/amoghghadge/cartquestandroid/ui/auth/LoginViewModel.kt
package com.amoghghadge.cartquestandroid.ui.auth

import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.amoghghadge.cartquestandroid.data.repository.UserRepository
import com.google.firebase.auth.FirebaseUser
import com.google.firebase.auth.GoogleAuthProvider
import com.google.firebase.auth.UserProfileChangeRequest
import com.google.firebase.auth.ktx.auth
import com.google.firebase.ktx.Firebase
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.launch
import kotlinx.coroutines.tasks.await

sealed class AuthState {
    object Unauthenticated : AuthState()
    object Loading : AuthState()
    data class Authenticated(val user: FirebaseUser) : AuthState()
    data class Error(val message: String) : AuthState()
}

class LoginViewModel : ViewModel() {
    private val auth = Firebase.auth
    private val userRepository = UserRepository()

    private val _authState = MutableStateFlow<AuthState>(
        if (auth.currentUser != null) AuthState.Authenticated(auth.currentUser!!)
        else AuthState.Unauthenticated
    )
    val authState: StateFlow<AuthState> = _authState.asStateFlow()

    fun signInWithEmail(email: String, password: String) {
        _authState.value = AuthState.Loading
        viewModelScope.launch {
            try {
                auth.signInWithEmailAndPassword(email, password).await()
                _authState.value = AuthState.Authenticated(auth.currentUser!!)
            } catch (e: Exception) {
                _authState.value = AuthState.Error(e.message ?: "Sign in failed")
            }
        }
    }

    fun signUpWithEmail(email: String, password: String, displayName: String) {
        _authState.value = AuthState.Loading
        viewModelScope.launch {
            try {
                auth.createUserWithEmailAndPassword(email, password).await()
                val user = auth.currentUser!!
                val profileUpdates = UserProfileChangeRequest.Builder()
                    .setDisplayName(displayName)
                    .build()
                user.updateProfile(profileUpdates).await()
                userRepository.createUserDocument(user)
                _authState.value = AuthState.Authenticated(user)
            } catch (e: Exception) {
                _authState.value = AuthState.Error(e.message ?: "Sign up failed")
            }
        }
    }

    fun signInWithGoogle(idToken: String) {
        _authState.value = AuthState.Loading
        viewModelScope.launch {
            try {
                val credential = GoogleAuthProvider.getCredential(idToken, null)
                auth.signInWithCredential(credential).await()
                val user = auth.currentUser!!
                userRepository.createUserDocument(user)
                _authState.value = AuthState.Authenticated(user)
            } catch (e: Exception) {
                _authState.value = AuthState.Error(e.message ?: "Google sign in failed")
            }
        }
    }

    fun signOut() {
        auth.signOut()
        _authState.value = AuthState.Unauthenticated
    }
}
